middle_end/flambda/simplify/simplify_common.ml

```diff
@@ -50,6 +50,49 @@ type simplify_toplevel =
   -> exn_cont_scope:Scope.t
   -> Expr.t * Upwards_acc.t
 
+let is_self_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.closure_info denv with
+  | Not_in_a_closure -> false
+  | Closure { code_id = fun_code_id;
+              return_continuation = _; exn_continuation = _; } ->
+    begin match Apply.call_kind apply with
+    | Function (Direct { code_id = apply_code_id; _ }) ->
+      Code_id.equal fun_code_id apply_code_id
+    | Method _ | C_call _
+    | Function (Indirect_known_arity _ | Indirect_unknown_arity)
+      -> false
+    end
+  (* CR: This case should not happend, but it's relatively safe to
+     return false here *)
+  | In_a_set_of_closure_but_not_yet_in_a_specific_closure -> false
+
+let is_tail_call dacc apply =
+  let denv = DA.denv dacc in
+  match DE.closure_info denv with
+  | Not_in_a_closure -> false
+  | In_a_set_of_closure_but_not_yet_in_a_specific_closure ->
+    (* same as for {!is_self_call} *) false
+  | Closure { code_id = _;
+              return_continuation = fun_cont;
+              exn_continuation = fun_exn_cont; } ->
+    (* 1st check: exn continuations match *)
+    let apply_exn_cont = Apply.exn_continuation apply in
+    Exn_continuation.equal fun_exn_cont apply_exn_cont &&
+    (* 2nd check: return continuations match *)
+    begin match Apply.continuation apply with
+    (* a function that raises unconditionally can be ta tail-call *)
+    | Never_returns -> true
+    | Return apply_cont -> Continuation.equal fun_cont apply_cont
+    end &&
+    (* 3rd match: for now, let's consider only function calls can be
+       tail-calls (C-calls have a different calling convention, and
+       methods perform a lot things) *)
+    begin match Apply.call_kind apply with
```

Comment:
I would probably replace the part in the brackets with just "for simplicity".  The issues are probably a bit tricky, and in any case we don't actually need this to work for C calls and methods, right?

Reply:
There is now a single function for detecting seIf tail calls, so I changed the comment and code for the 3rd check (since we only detect self calls, it implies that only ocaml calls are considered).